drivers: rtc: nucleo_wb55rg: Fix RTC callback issue when using power management config

krisnaresi opened the pull request:
It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the `DBP` bit on `PWR control register 1` via `LL_PWR_EnableBkUpAccess` function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG.

Fixes: #68673

ajarmouni-st commented:
@krisnaresi You have compliance issues in CI.

```
Commit 94f65adbd8:
1: UC5 Commit title exceeds max length (95>75): "drivers: rtc: nucleo_wb55rg: Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
3: UC4 Commit message body line exceeds max length (404>75): "It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the DBP bit on PWR control register 1 via LL_PWR_EnableBkUpAccess function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG."

Commit effdc[29](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:30)b77:
1: UC2 Commit message does not contain a 'Signed-off-by:' line
1: UC3 Commit title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
1: UC6 Commit message body is empty, should at least have 1 line(s).
3: B6 Body message is missing
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

 94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
 effdc29b77eedeb61679e5ce3de8b13d8b45[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:39)c3: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
effdc29b77eedeb61679e5ce3de8b13d8b4538c3: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
```

krisnaresi commented:
> @krisnaresi You have compliance issues in CI.
> 
> ```
> Commit 94f65adbd8:
> 1: UC5 Commit title exceeds max length (95>75): "drivers: rtc: nucleo_wb55rg: Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
> 3: UC4 Commit message body line exceeds max length (404>75): "It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the DBP bit on PWR control register 1 via LL_PWR_EnableBkUpAccess function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG."
> 
> Commit effdc[29](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:30)b77:
> 1: UC2 Commit message does not contain a 'Signed-off-by:' line
> 1: UC3 Commit title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
> 1: UC6 Commit message body is empty, should at least have 1 line(s).
> 3: B6 Body message is missing
> Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details
> 
>  94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
> 94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
>  effdc29b77eedeb61679e5ce3de8b13d8b45[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:39)c3: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
> effdc29b77eedeb61679e5ce3de8b13d8b4538c3: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
> ```

@ajarmouni-st, I think I have resolve commit message. Still waiting for re-run on the workflow. Let me know if you spot something. Thanks.

krisnaresi committed:
drivers: rtc: Fix RTC alarm when using both CONFIG_COUNTER and CONFIG_PM

It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently,
the RTC alarm callback can be used only once (in some cases, it just
won't work at all, e.g., using CONFIG_BT). By set the DBP bit on PWR
control register 1 via LL_PWR_EnableBkUpAccess function to temporarily
disable write protection every time we assign RTC alarm, we can register
alarm callback correctly. Tested on Nucleo WB55RG.

Fixes: #68673
Signed-off-by: Krisna Resi <krisna_resi@ymail.com>

ajarmouni-st commented:
The good news, I run `tests/drivers/counter/counter_basic_api/drivers.counter.basic_api` with `CONFIG_PM=y` & all the tests pass with this patch. Previously the tests `test_single_shot_alarm_notop` & `test_all_channels` were failing for RTC counter.

So it would be great if you can add a commit with the following test case in `tests/drivers/counter/counter_basic_api/testcase.yaml`
```
  drivers.counter.basic_api.stm32_pm:
    tags:
      - drivers
      - counter
      - power
    depends_on: counter
    min_ram: 16
    platform_allow:
      - nucleo_wb55rg
    extra_configs:
      - CONFIG_PM=y
    timeout: 600
```

ajarmouni-st commented:
Update: `test_single_shot_alarm_notop` is still failing with `CONFIG_PM=y` ...

krisnaresi commented:
Hi @ajarmouni-st, this morning I already retest on my project with the exact main branch commit [b47c66e](https://github.com/krisnaresi/zephyr/commit/b47c66e3f293c171e388fb285ae4f30e31068e62) and make changes as per this pull request. Tested using source code that I shared with [you](https://github.com/zephyrproject-rtos/zephyr/issues/68673#issuecomment-1945864671) and with those [prj.conf](https://github.com/zephyrproject-rtos/zephyr/issues/68673#issuecomment-1972473821). It's work properly on my side.

Now, I don't know how to do ZTest, maybe you can share the step and build command on how to reproduce `test_single_shot_alarm_notop` error to me? And I see there is [overlay](https://github.com/krisnaresi/zephyr/blob/main/tests/drivers/counter/counter_basic_api/boards/nucleo_wb55rg.overlay) here with a lot of timers, do you think it can cause this problem?

ajarmouni-st commented:
> Now, I don't know how to do ZTest, maybe you can share the step and build command on how to reproduce `test_single_shot_alarm_notop` error to me? And I see there is [overlay](https://github.com/krisnaresi/zephyr/blob/main/tests/drivers/counter/counter_basic_api/boards/nucleo_wb55rg.overlay) here with a lot of timers, do you think it can cause this problem?

You can run the test with
`west build -p always -b nucleo_wb55rg -T tests/drivers/counter/counter_basic_api/drivers.counter.basic_api -- -DCONFIG_PM=y`

In any case, since the RTC counter tests are not all passing with `CONFIG_PM=y`, adding a test case for it isn't possible right now & can be left for another PR when the problem is fixed.

krisnaresi commented:
Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.

I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.

erwango commented:
> Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.
> 
> I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.

One last idea to test would be to align implementation on driver/rtc/rtc_stm32.c which seem more complete from BkUp Access point of view. If this doesn't help solving the last isue, I don't mind merging as is, as this fixes some bugs at least.

ajarmouni-st commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -240,6 +240,12 @@ tick_t rtc_stm32_read(const struct device *dev)
 #endif /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS */
 	ARG_UNUSED(dev);
 
+	/* Enable Backup access */
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || \
+	defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+	LL_PWR_EnableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
```
@krisnaresi I'am ready to approve the PR, just need this change to be taken into account. Thanks for your collaboration!

krisnaresi replied:
@ajarmouni-st, I think you're correct, let me push new commit for that. Thanks.

krisnaresi commented:
> > Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.
> > I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.
> 
> One last idea to test would be to align implementation on driver/rtc/rtc_stm32.c which seem more complete from BkUp Access point of view. If this doesn't help solving the last isue, I don't mind merging as is, as this fixes some bugs at least.

Hi @erwango, I already implement `LL_PWR_EnableBkUpAccess()` on `rtc_stm32_get_time` function on `drivers/rtc/rtc_ll_stm32.c`, but `test_single_shot_alarm_notop` test still failed. As I recall, `rtc_ll_stm32.c` still not implement alarm yet, so in my opinion, I don't think it can solve any test alarm problem.

ajarmouni-st commented:
@krisnaresi no I just run the tests w/t this change, they pass. No need to add anything.

krisnaresi commented:
> @krisnaresi no I just run the tests w/t this change, they pass. No need to add anything.

Oh, ok then. So, it's ready to merge I suppose? I'm not sure if I should be the one to merge it or somebody else that have write access..

ajarmouni-st commented:
> Oh, ok then. So, it's ready to merge I suppose? 

Pending a second approval

> somebody else that have write access.

Yes
